packages/node-core/src/indexer/project.service.ts

```diff
@@ -260,31 +260,65 @@ export abstract class BaseProjectService<API extends IApi, DS extends BaseDataSo
+    return dataSources.some((ds) => ds.endBlock === undefined || ds.endBlock > height);
+  }
+
   // This gets used when indexing blocks, it needs to be async to ensure dynamicDs is updated within workers
   async getDataSources(blockHeight?: number): Promise<DS[]> {
     const dataSources = this.project.dataSources;
     const dynamicDs = await this.dynamicDsService.getDynamicDatasources();
 
     return [...dataSources, ...dynamicDs].filter(
-      (ds) => blockHeight === undefined || (ds.startBlock !== undefined && ds.startBlock <= blockHeight)
+      (ds) =>
+        blockHeight === undefined ||
+        (ds.startBlock !== undefined &&
+          ds.startBlock <= blockHeight &&
+          (ds.endBlock === undefined || ds.endBlock >= blockHeight))
     );
   }
 
   getDataSourcesMap(): BlockHeightMap<DS[]> {
-    assert(isMainThread, 'This method is only avaiable on the main thread');
+    assert(isMainThread, 'This method is only available on the main thread');
     const dynamicDs = this.dynamicDsService.dynamicDatasources;
-
     const dsMap = new Map<number, DS[]>();
+    const activeDataSources = new Set<DS>();
+    //events denote addition or deletion of datasources from height-datasource map entries at the specified block height
+    const events: {
+      block: number; //block height at which addition or deletion of datasource should take place
+      start: boolean; //if start=TRUE, add the datasource. otherwise remove the datasource.
+      ds: DS;
+    }[] = [];
 
-    // Loop through all projects
     for (const [height, project] of this.projectUpgradeService.projects) {
-      // Iterate all the DS at the project height
-      [...project.dataSources, ...dynamicDs]
-        .filter((ds): ds is DS & {startBlock: number} => !!ds.startBlock)
-        .sort((a, b) => a.startBlock - b.startBlock)
-        .forEach((ds, index, dataSources) => {
-          dsMap.set(Math.max(height, ds.startBlock), dataSources.slice(0, index + 1));
-        });
+      const dataSources = [...project.dataSources, ...dynamicDs].filter(
+        (ds): ds is DS & {startBlock: number} => !!ds.startBlock
+      );
+
+      dataSources.forEach((ds) => {
+        const startBlock = Math.max(height, ds.startBlock);
+
+        //ds is to be added at startBlock
+        events.push({block: startBlock, start: true, ds});
+        if (ds.endBlock) {
+          const endBlock = ds.endBlock + 1;
+          //ds is to be removed at endBlock
+          events.push({block: endBlock, start: false, ds});
+        }
+      });
+    }
+
+    // sort events by block in ascending order, start events come before end events
+    events.sort((a, b) => a.block - b.block || Number(b.start) - Number(a.start));
+
+    for (const event of events) {
+      if (event.start) {
+        activeDataSources.add(event.ds);
+      } else {
+        activeDataSources.delete(event.ds);
+      }
+      dsMap.set(event.block, Array.from(activeDataSources));
```

Comment:
I think this needs to be done in a single loop. An issue I can see is datasources being mixed between project versions. It would also be much simpler